frame/system/src/offchain.rs

```diff
@@ -17,351 +17,486 @@
-			local_accounts_and_keys.sort_by(|a, b| a.0.cmp(&b.0));
-
-			// get all the matching accounts
-			accounts.into_iter().filter_map(|acc| {
-				let idx = local_accounts_and_keys.binary_search_by(|a| a.0.cmp(&acc)).ok()?;
-				local_accounts_and_keys.get(idx).cloned()
-			}).collect()
-		} else {
-			// just return all account ids and keys
-			local_accounts_and_keys
+	fn clone(&self) -> Self {
+		Self {
+			index: self.index,
+			id: self.id.clone(),
+			public: self.public.clone(),
 		}
 	}
+}
 
-	fn can_sign_with(accounts: Option<impl IntoIterator<Item = T::AccountId>>) -> bool {
-		// early exit if we care about any account.
-		if accounts.is_none() {
-			!S::all().is_empty()
-		} else {
-			!Self::find_local_keys(accounts).is_empty()
+/// App specific crypto trait that provides sign/verify
+/// abilities to offchain workers. Implementations of this
+/// trait should specify the app-specific public/signature
+/// types.
+pub trait AppCrypto<Public, Signature> {
+	type RuntimeAppPublic: RuntimeAppPublic;
+	// TODO [ToDr] The conversions are messy, clean them up.
+	//
+	// The idea would be to have some implementation for `RuntimeAppPublic`
+	// to convert to and from generic types.
+	// Maybe even a method like:
+	// impl RuntimeAppPublic {
+	//  fn into_public<T: From<Self::Generic>>(&self) -> T;
+	// }
+	// so an ability to convert the runtime app public into
+	// some type that is reachable from the inner (wrapped) generic
+	// crypto type.
+	// So example:
+	// ImOnline(sr25519) = RuntimeAppPublic
+	// sr25519 = Generic
+	// MutliSigner = From<sr25519>
+	type GenericPublic:
+		From<Self::RuntimeAppPublic>
+		+ Into<Self::RuntimeAppPublic>
+		+ TryFrom<Public>
+		+ Into<Public>;
+	type GenericSignature:
+		From<<Self::RuntimeAppPublic as RuntimeAppPublic>::Signature>
+		+ Into<<Self::RuntimeAppPublic as RuntimeAppPublic>::Signature>
+		+ TryFrom<Signature>
+		+ Into<Signature>;
+
+	fn sign(payload: &[u8], public: Public) -> Option<Signature> {
+		let p: Self::GenericPublic = public.try_into().ok()?;
+		let x = Into::<Self::RuntimeAppPublic>::into(p);
+		x.sign(&payload)
+			.map(|x| {
+				let sig: Self::GenericSignature = x.into();
+				sig
+			})
+			.map(Into::into)
+	}
+
+	fn verify(payload: &[u8], public: Public, signature: Signature) -> bool {
+		let p: Self::GenericPublic = match public.try_into() {
+			Ok(a) => a,
+			_ => return false
+		};
+		let x = Into::<Self::RuntimeAppPublic>::into(p);
+		let signature: Self::GenericSignature = match signature.try_into() {
+			Ok(a) => a,
+			_ => return false
+		};
+		let signature = Into::<<
+			Self::RuntimeAppPublic as RuntimeAppPublic
+		>::Signature>::into(signature);
+
+		x.verify(&payload, &signature)
+	}
+
+}
+
+/// A wrapper around the types which are used for signing transactions.
+/// This trait should be implemented on the runtime.
+pub trait SigningTypes: crate::Trait {
+	// TODO [ToDr] Could this be just `T::Signature as traits::Verify>::Signer`?
```

Comment:
Same here, either remove or create an issue and add a number for reference (fine with single issue with a checklist of things to address).